bootstrap/bpdoc/reader.go

```diff
@@ -0,0 +1,511 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bpdoc
+
+import (
+	"fmt"
+	"go/ast"
+	"go/doc"
+	"go/parser"
+	"go/token"
+	"html/template"
+	"reflect"
+	"runtime"
+	"strconv"
+	"strings"
+	"sync"
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/google/blueprint/proptools"
+)
+
+type Reader struct {
+	pkgFiles map[string][]string // Map of package name to source files, provided by constructor
+
+	mutex sync.Mutex
+	pkgs  map[string]*doc.Package    // Map of package name to parsed Go AST, protected by mutex
+	ps    map[string]*PropertyStruct // Map of type name to property struct, protected by mutex
+}
+
+func NewReader(pkgFiles map[string][]string) *Reader {
+	return &Reader{
+		pkgFiles: pkgFiles,
+		pkgs:     make(map[string]*doc.Package),
+		ps:       make(map[string]*PropertyStruct),
+	}
+}
+
+func funcToPkgPath(f string) (string, error) {
+	p := strings.Split(f, "/")
+
+	pkgPath := strings.Join(p[:len(p)-1], "/")
+
+	p = strings.Split(p[len(p)-1], ".")
+	if len(p) < 2 {
+		return "", fmt.Errorf("failed to extract package path from %q", f)
+	}
+
+	if pkgPath != "" {
+		pkgPath += "/"
+	}
+	pkgPath += p[0]
+
+	return pkgPath, nil
+}
+
+func (r *Reader) ModuleType(name string, factory reflect.Value) (*ModuleType, error) {
+	f := runtime.FuncForPC(factory.Pointer())
+
+	pkgPath, err := funcToPkgPath(f.Name())
+	if err != nil {
+		return nil, err
+	}
+
+	factoryName := strings.TrimPrefix(f.Name(), pkgPath+".")
+
+	text, err := r.moduleTypeDocs(pkgPath, factoryName)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ModuleType{
+		Name:    name,
+		PkgPath: pkgPath,
+		Text:    text,
+	}, nil
+}
+
+func (r *Reader) Package(pkgPath string) (*Package, error) {
+	pkg, err := r.pkg(pkgPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Package{
+		Name:    pkg.Name,
+		PkgPath: pkgPath,
+		Text:    pkg.Doc,
+	}, nil
+}
+
+func (r *Reader) pkgDocs(pkgPath string) (string, error) {
+	pkg, err := r.pkg(pkgPath)
+	if err != nil {
+		return "", err
+	}
+
+	return pkg.Doc, nil
+}
+
+func (r *Reader) moduleTypeDocs(pkgPath, name string) (string, error) {
+	pkg, err := r.pkg(pkgPath)
+	if err != nil {
+		return "", err
+	}
+
+	for _, fn := range pkg.Funcs {
+		if fn.Name == name {
+			return fn.Doc, nil
+		}
+	}
+
+	// The doc package may associate the method with the type it returns, so iterate through those too
+	for _, typ := range pkg.Types {
+		for _, fn := range typ.Funcs {
+			if fn.Name == name {
+				return fn.Doc, nil
+			}
+		}
+	}
+
+	return "", nil
+}
+
+// Return the PropertyStruct associated with a property struct type.  The type should be in the
+// format <package path>.<type name>
+func (r *Reader) PropertyStruct(pkgPath, name string, defaults reflect.Value) (*PropertyStruct, error) {
```

Comment:
I have a similar feeling toward PropertyStruct-related functions, but have no clear idea how to better organize them. Would be willing to rearrange them myself once getting familiar with the code base, though.